flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/DataSetConversionUtil.java

```diff
@@ -0,0 +1,172 @@
+	 *
+	 * @param sessionId the sessionId of {@link MLEnvironmentFactory}
+	 * @param table the Table to convert.
+	 * @return the converted DataSet.
+	 */
+	public static DataSet <Row> fromTable(Long sessionId, Table table) {
+		return MLEnvironmentFactory
+			.get(sessionId)
+			.getBatchTableEnvironment()
+			.toDataSet(table, Row.class);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified TableSchema.
+	 *
+	 * @param sessionId the sessionId of {@link MLEnvironmentFactory}
+	 * @param data   the DataSet to convert.
+	 * @param schema the specified TableSchema.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, TableSchema schema) {
+		return toTable(sessionId, data, schema.getFieldNames(), schema.getFieldTypes());
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames and colTypes.
+	 *
+	 * @param sessionId sessionId the sessionId of {@link MLEnvironmentFactory}.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @param colTypes the specified colTypes. This variable is used only when the
+	 *                 DataSet is produced by a function and Flink cannot determine
+	 *                 automatically what the produced type is.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, String[] colNames, TypeInformation <?>[] colTypes) {
+		return toTable(MLEnvironmentFactory.get(sessionId), data, colNames, colTypes);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames.
+	 *
+	 * @param sessionId sessionId the sessionId of {@link MLEnvironmentFactory}.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, String[] colNames) {
+		return toTable(MLEnvironmentFactory.get(sessionId), data, colNames);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames and colTypes.
+	 *
+	 * @param session the MLEnvironment using to convert DataSet to Table.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @param colTypes the specified colTypes. This variable is used only when the
+	 *                 DataSet is produced by a function and Flink cannot determine
+	 *                 automatically what the produced type is.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(MLEnvironment session, DataSet <Row> data, String[] colNames, TypeInformation <?>[] colTypes) {
+		try {
+			// Try to add row type information for the dataset to be converted.
+			// In most case, this keeps us from the rolling back logic in the catch block,
+			// which adds an unnecessary map function just in order to add row type information.
+			if (data instanceof SingleInputUdfOperator) {
```

Comment:
After improving the test I found that this piece of heuristic improvement logic doesn't work.
I actually go ahead and removed this in https://github.com/walterddr/flink/commit/c55dfd899d17fb59e62211ab2dd05ff82ae1480f. Please kindly take a look

Reply:
Hi, I'm just confused about the meaning of 'not work', could you explain it a bit more in detail?
I looked into the case `testE2E`, for the invocation on `table2`, our heuristic prevented it from falling back to the explicit type definition. You can see that if you remove the heuristic. But for `tableFromDataStreamWithTableSchema`, it fell to exception branch because the `input` dataset was reused and the `getType` API was accessed twice as your described in comments.

Or maybe I just missed something ……^_^

Reply:
It is very easy to reproduce, if you checkout the https://github.com/walterddr/flink/commit/c55dfd899d17fb59e62211ab2dd05ff82ae1480f and adding line 105-112 back. you would see the testE2E fails - interestingly the test failure is actually occurring in the 3rd `toTable` call, not the second one. please see https://github.com/walterddr/flink/commit/c55dfd899d17fb59e62211ab2dd05ff82ae1480f#r36615861

I am not sure whether the problem is with the test or with the logic itself (I copied the E2E test from the `DataStreamConversionUtilsTest` and I think we should expect feature parity between the two).


It could've been a mistake in the test case itself. Please kindly take a look and see if this is suppose to be a test problem or a conversion logic problem. if (1) test problem, then we need to fix the test and simultaneously fix the heuristic in the `DataStreamConversionUtils`; (2) if it is in fact a conversion problem we should remove the heuristic in both Utils. Please let me know .

Reply:
There're two lines confuse me in `testE2E`, 
  1. The line `thrown.expect(ValidationException.class);` , why is an `ValidationException` expected?
  2. The line `env.execute();` seems have nothing to execute because of the previous `print()` ,am I right?

So I just comment them out, add the heuristic lines back and move ahead. I do reproduce the exception you mentioned, in the 3rd `toTable` invocation. I think that's because type information has been added to `input` in the 2nd invocation already. Flink complains about this when the same thing applied to the identical `input` object in the 3rd invocation. So I think at least it proves that our heuristic works. And it also reveals a corner case we missed, in which an identical `DataSet` object is passed to `toTable` more than once.  UT will pass after adding `IllegalStateException` to catch block. The new `toTable` mothod now looks like:
```java
	public static Table toTable(MLEnvironment session, DataSet <Row> data, String[] colNames, TypeInformation <?>[] colTypes) {
		try {
			if (data instanceof SingleInputUdfOperator) {
				((SingleInputUdfOperator) data).returns(new RowTypeInfo(colTypes, colNames));
			} else if (data instanceof TwoInputUdfOperator) {
				((TwoInputUdfOperator) data).returns(new RowTypeInfo(colTypes, colNames));
			}
			return toTable(session, data, colNames);
		} catch (ValidationException | IllegalStateException ex) {
			// currently ValidationException will be thrown and caught for further processing.
			// Because the getType() API of the Transformation can only be accessed once.
			// This can be improve if we add "isTypeSet()" API to the Transformation class.
			if (null == colTypes) {
				throw ex;
			} else {
				DataSet <Row> t = fallbackToExplicitTypeDefine(data, colNames, colTypes);
				return toTable(session, t, colNames);
			}
		}
```

I just wanna make sure it ok to add another exception to the origin anti-pattern code ^_^ ..

Reply:
I think the 3rd `toTable` was NOT suppose to work at all (e.g. the same DataStream/DataSet object should not be converted to table twice)

If so I think we should even remove the 3rd `toTable` operation all together instead of adding another catcher - especially because `IllegalStateException` is a `java.lang` exception instead of a Flink defined exception, there might be other ways it was thrown.

Reply:
> If so I think we should even remove the 3rd toTable operation all together instead of adding another catcher - especially because IllegalStateException is a java.lang exception instead of a Flink defined exception, there might be other ways it was thrown.

make sence.